fix: Improve error for missing (Translate)CustomTagBlock template #35829

qasimgulzar opened the pull request:
## Description

Under this PR I have fixed a open issue [35829](https://github.com/openedx/edx-platform/issues/35829). I have placed checks and exception handlers to gracefully inform the users.

## Supporting information

https://tasks.opencraft.com/browse/BB-9666

## Testing instructions

1. Check out [master/main](qasimgulzar:qasim/35829) on edx-platform allow CMS and LMS to reload.
2. Goto authering frontend.
3. Add new unit
4. click on `Advanced`
5. Select customtag/book (currently in master after adding empty customtag xblock it will show you error/500)
6. Verify you should be able to update customtag xml, you should also be able to remove this problem.
7. Screenshots and recording is attached for your reference

**Note**

I have used course attach to the [GITHUB ISSUE](https://github.com/openedx/edx-platform/issues/35829#issuecomment-2471198229) for testing.


<img width="1281" height="1247" alt="image" src="https://github.com/user-attachments/assets/b84fa0fd-19ea-4dcf-81b3-a4ddf60da34a" />

**Customtag with missing template.**
<img width="1528" height="812" alt="image" src="https://github.com/user-attachments/assets/142f9d52-bca7-4776-a54e-c7be16510427" />

Commit:
fix: Improve error for missing (Translate)CustomTagBlock template #35829

openedx-webhooks commented:
Thanks for the pull request, @qasimgulzar!


This repository is currently maintained by `@openedx/wg-maintenance-edx-platform`.


Once you've gone through the following steps feel free to tag them in a comment and let them know that your changes are ready for engineering review.

<details><summary>:radio_button: Get product approval</summary>

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.
</details>

<details><summary>:radio_button: Provide context</summary>

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads
</details>





<details><summary>:radio_button: Get a green build</summary>

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.
</details><details>


---

<details><summary>Where can I find more information?</summary>

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)
</details>

<details><summary>When can I expect my changes to be merged?</summary>

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*
</details>
<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

Commit:
fix: TranslateCustomTagBlock replace tag with customtag element

qasimgulzar commented:
<img width="1647" height="1074" alt="image" src="https://github.com/user-attachments/assets/ca313865-5d8a-42cc-879a-38586bddc831" />

qasimgulzar commented:
https://github.com/user-attachments/assets/8ea48dcc-220f-4a54-ba17-43b82551a8fe

Commit:
fix: TranslateCustomTagBlock replace tag with customtag element

Commit:
fix: TranslateCustomTagBlock replace tag with customtag element

qasimgulzar commented:
## Translated Tag Test

<img width="1557" height="840" alt="image" src="https://github.com/user-attachments/assets/fab99136-7154-499f-86d5-b487ac51043e" />
<img width="1313" height="331" alt="image" src="https://github.com/user-attachments/assets/4adeac98-c68a-4bf5-bb30-244b30a7b4b1" />

qasimgulzar commented:
## Custom Tag Test

<img width="1530" height="953" alt="image" src="https://github.com/user-attachments/assets/fda9fc4b-e9e9-469f-b03b-cdd41387be4c" />
<img width="1342" height="257" alt="image" src="https://github.com/user-attachments/assets/cc00df23-940e-46b0-ab7e-a0614109b6a0" />

tecoholic commented on xmodule/template_block.py:
```diff
@@ -76,6 +73,8 @@ def studio_view(self, _context):
     def render_template(self, system, xml_data):
         '''Render the template, given the definition xml_data'''
+        if not xml_data:
+            return "Please set the template for this custom tag."
```
@qasimgulzar Hi, I am not sure how to test this bit. Can you kindly add a step by step guide for testing the various scenarios? For reference, see the testing instructions in https://github.com/openedx/edx-platform/pull/32210

qasimgulzar replied:
@tecoholic I have updated the testing instructions. There is a recording attached please give it look as well.

tecoholic replied:
@qasimgulzar Thak you for the testing instructions. That helped. On the related note, I ran into errors when the template is empty, i.e. `impl=""`. I think the issue is do to with the `if child_impl is not None` in line 84, which check specifically for the missing `impl` attribute, but considers an empty string to be valid template name. Can you kindly handle that?

qasimgulzar replied:
@tecoholic great catch, let me look into it.

qasimgulzar replied:
@tecoholic I have updated the PR

tecoholic commented on xmodule/template_block.py:
```diff
@@ -86,16 +85,19 @@ def render_template(self, system, xml_data):
+            template_block_data = template_block.data
+        except ItemNotFoundError as ex:
+            log.exception(f"Could not find template block for custom tag with Id {template_name}")
+            template_block_data = f"Could not find template block for custom tag with Id {template_name}. Error: {ex}"
```
* The strings here seem identical. It would be better to construct this once and use it both places. The "Error: " bit in the second one renders as `Error: BlockKey(type='custom_tag_template', id='mytag3')` for the instructior, which is not very helpful, as it just repeats the same "mytag3" missing tag.
* Here the exception is a known one. Can we drop the log.exception()? Or do you see any value is adding this to the server logs?

qasimgulzar replied:
@tecoholic I replaced the exception with info, but we atleast need to track it in logs so if someone has to debug this  from logs they can atleast. 

I also updated the message.

tecoholic commented on xmodule/template_block.py:
```diff
@@ -120,28 +122,28 @@ def export_to_file(self):
         node.attrib['impl'] = tag
-
-        return runtime.process_xml(etree.tostring(node))
+        return node
```
I don't understand the refactoring here. Can you kindly explain it? The complete change of implementation and dropping of `runtime.error_tracker` is concerning for me. But I don't have a good understanding of this code base to know what's safe and not.

qasimgulzar replied:
@tecoholic as per old implementation `TranslateCustomTagBlock` was supposed to replace the outermost xml node with `customtag` however this implementation was broken. So to fix the implementation I have updated this class.

qasimgulzar replied:
https://drive.google.com/file/d/1mbYcu6gUfrmvmg9P167fZSjj0b9dnEo7/view?usp=sharing

@tecoholic I have recorded a video for you to explain the solution.

qasimgulzar replied:
this will also fix [this](https://github.com/openedx-unsupported/openedx-test-course/issues/16) issue

inorder to being more descriptive, here class DocString is explaining the intention `Converts olx of the form `<$custom_tag attr="" attr=""/>` to CustomTagBlock`

So it is missing part of the implementation which is causing the issues.

<img width="826" height="259" alt="image" src="https://github.com/user-attachments/assets/781f4eed-c676-4abf-bbed-6d7f18e138d6" />

tecoholic replied:
@qasimgulzar Thank you. I tested with and without changes to the `TranslateCustomBlock` and see that without your refactoring, the entire Vertical Block fails to render.

Also, I edited the OLX to include a `<book>` tag in the vertical and it works without issues. So both `<customtag impl="book".../>` and `<book ..../>` works.

Nice job getting this sorted.

Commit:
fix: PR comments accommodated

qasimgulzar commented:
>@qasimgulzar Thak you for the testing instructions. That helped. On the related note, I ran into errors when the template is empty, i.e. impl="". I think the issue is do to with the if child_impl is not None in line 84, which check specifically for the missing impl attribute, but considers an empty string to be valid template name. Can you kindly handle that?

@tecoholic could you please add few more details? about the steps to reproduce this issue?

Commit:
fix: added fix when impl=""

Author:    qasimgulzar <qasim@primzel.com>

Revert "fix: added fix when impl="""

This reverts commit 5f8d4cded5c73b75783252f24d8dd19e2cd6454a.

fix: added fix when impl=""

qasimgulzar commented:
@tecoholic I have figured it out and pushed a fix in last commit.

tecoholic reviewed (approved):
@qasimgulzar :+1:

- [x] I tested this: Verified the custom tags in the OLX are rendered without issues and shows relevant error messages.
- [x] I read through the code
- [x] I checked for accessibility issues
- [ ] Includes documentation

tecoholic commented:
@kdmccormick Pinging you for review as you have the best context for CustomTagBlock (https://github.com/openedx/edx-platform/issues/35829).

kdmccormick commented:
Thank you!
